pkg/operator/operator.go

```diff
@@ -229,76 +257,204 @@ func (k *Kontroller) withLeaderElection() error {
+	err := k.checkPostRebootActions()
 	if err != nil {
-		glog.Infof("Failed listing nodes %v", err)
+		glog.Errorf("Failed to check post-reboot actions: %v", err)
 		return
 	}
 
-	glog.V(6).Infof("Found nodes: %+v", nodelist.Items)
+	// find nodes which just rebooted but aren't yet running post-reboot actions.
+	// remove any old action annotations and add the reboot-action=post label.
+	// this starts the post-reboot actions.
+	err = k.markPostRebootActionable()
+	if err != nil {
+		glog.Errorf("Failed to update recently rebooted nodes: %v", err)
+		return
+	}
 
-	justRebootedNodes := k8sutil.FilterNodesByAnnotation(nodelist.Items, justRebootedSelector)
+	// find nodes with reboot-action=pre label and check if all actions are complete.
+	// if all actions are completed successfully then remove the reboot-action=pre label
+	// and set reboot-ok=true, telling the agent it's time to reboot.
+	err = k.checkPreRebootActions()
+	if err != nil {
+		glog.Errorf("Failed to check pre-reboot actions: %v", err)
+		return
+	}
 
-	if len(justRebootedNodes) > 0 {
-		glog.Infof("Found %d rebooted nodes, setting annotation %q to false", len(justRebootedNodes), constants.AnnotationOkToReboot)
+	// take some number of the rebootable nodes.
+	// remove any old action annotations and add the reboot-action=pre label.
+	// this starts the pre-reboot actions.
+	err = k.markPreRebootActionable()
+	if err != nil {
+		glog.Errorf("Failed to update rebootable nodes")
+		return
 	}
+}
 
-	for _, n := range justRebootedNodes {
-		glog.Infof("Setting 'ok-to-reboot=false' for %q", n.Name)
-		if err := k8sutil.SetNodeAnnotations(k.nc, n.Name, map[string]string{
-			constants.AnnotationOkToReboot: constants.False,
-		}); err != nil {
-			glog.Infof("Failed setting annotation %q on node %q to false: %v", constants.AnnotationOkToReboot, n.Name, err)
+func (k *Kontroller) checkPreRebootActions() error {
+	nodelist, err := k.nc.List(v1meta.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("Failed listing nodes %v", err)
+	}
+
+	preRebootNodes := k8sutil.FilterNodesByLabel(nodelist.Items, preRebootSelector)
+
+	for _, n := range preRebootNodes {
+		res := action.ActionsCompleted(n, k.preRebootActions)
+		if res == action.ActionsSucceeded {
+			glog.Infof("Deleting label %q for %q", constants.LabelRebootAction, n.Name)
+			err := k8sutil.DeleteNodeLabels(k.nc, n.Name, []string{constants.LabelRebootAction})
+			if err != nil {
+				glog.Infof("Failed deleting label %q on node %q: %v", constants.LabelRebootAction, n.Name, err)
+				continue
+			}
+
+			// manually call UpdateNodeRetry so that we can make sure the node still wants to reboot
+			glog.Infof("Marking %q ok to reboot", n.Name)
+			err = k8sutil.UpdateNodeRetry(k.nc, n.Name, func(node *v1api.Node) {
+				if !wantsRebootSelector.Matches(fields.Set(node.Annotations)) {
+					glog.Warningf("Node %v no longer wanted to reboot while we were trying to mark it so: %v", node.Name, node.Annotations)
+					return
```

Comment:
This line is part of an anonymous function that starts on line 317. It's a little weird, but this node annotation update needs to make sure the node still wants to reboot before it gets marked to do so.